orb-ui/src/simulation.rs

```diff
@@ -1,90 +1,229 @@
+                // write 441 into /sys/class/gpio/export
+                if let Err(e) = std::fs::write("/sys/class/gpio/export", "441") {
+                    error!("Button cannot be configured, use `sudo echo 441 > /sys/class/gpio/export`: {}", e);
+                }
+                break;
+            }
+
+            loop {
+                // read cat /sys/class/gpio/PQ.06/value in a loop until it's 1
+                if let Ok(value) =
+                    std::fs::read_to_string("/sys/class/gpio/PQ.06/value")
+                {
+                    if value.trim() == "1" {
+                        break;
+                    }
+                    time::sleep(Duration::from_millis(100)).await;
+                }
+            }
+        } else if self_serve {
+            // waiting for app button press
+            time::sleep(Duration::from_secs(6)).await;
+        }
+        info!("Starting capture");
+
+        // biometric capture start, either:
+        // - cone button pressed, or
+        // - app button pressed
+        ui.signup_start();
+
+        let mut x_angle = 1_i32;
+        if showcar {
+            time::sleep(Duration::from_secs(2)).await;
+            let steps = 2000 / 30_u32; // 30ms per step, 200ms total
+            let gimbal_x_steps = 45000_u32 / steps;
+            for i in 0..steps {
+                ui.gimbal(gimbal_x_steps * i, 90000);
+                time::sleep(Duration::from_millis(30)).await;
+            }
+        }
+
+        // waiting for the user to be in correct position
+        ui.biometric_capture_distance(false);
+        time::sleep(Duration::from_millis(8500)).await;
+
+        let mut biometric_capture_error = false;
+
+        // 100 steps, 80ms per step, 8 seconds total
+        let biometric_capture_interval_ms = 70;
+        for i in 1..=100 {
+            if !showcar && (30..=50).contains(&i) {
+                // simulate user moving away
+                ui.biometric_capture_distance(false);
+                ui.biometric_capture_occlusion(true);
+            } else {
+                // capture is making progress
+                ui.biometric_capture_distance(true);
+                ui.biometric_capture_occlusion(false);
+                ui.biometric_capture_distance(true);
+                ui.biometric_capture_progress(i as f64 / 100.0);
+            }
+
+            // randomly simulate error
+            if !showcar && i == 50 && rand::random::<u8>() % 5 == 0 {
+                info!("⚠️ Simulating biometric capture error");
+                biometric_capture_error = true;
+                ui.signup_fail(SignupFailReason::Timeout);
+                break;
+            }
+
+            // simulate gimbal movement
+            let x_base = if i < 50 { 40000 } else { 35000 };
+            let x_rand = rand::random::<i32>() % 1000;
+            let y_rand = rand::random::<i32>() % 1000;
+            let x_sign: i32 = if rand::random::<u8>() % 2 == 0 { 1 } else { -1 };
```

Comment:
Probably more straightforward to do `rand::random::<bool>()`

Also, x_rand is already a random signed number. Whats the point of `x_sign` then?

Reply:
I thought the modulo wouldn't work for negative number, it does :)